ir/numbers.go

```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
 	}
 	panic("bug: unknown number type")
 }
+
+func unmarshalInt(p []byte, n *Int) error {
+	if string(p) == "null" {
+		return nil
+	}
+	var z big.Int
+	_, ok := z.SetString(strings.Replace(string(p), "\"", "", -1), 10)
+	if !ok {
+		return fmt.Errorf("not a valid big integer int: %s", p)
+	}
+	n.Int = &z
+	return nil
+}
+
+func unmarshalFloat(p []byte, n *Float) error {
+	if string(p) == "null" {
+		return nil
+	}
+	// Using precision 53 to accommodate for big.NewFloat which uses 53 preciosion by default.
+	// If we change this marshalling behavior to use MarshalText we need to change to 64 precision.
+	// Check: https://github.com/golang/go/issues/45309
+	z, _, err := big.ParseFloat(strings.Replace(string(p), "\"", "", -1), 10, 53, big.ToNearestEven)
+	if err != nil {
+		return err
+	}
+	n.Float = z
+	return nil
+}
+func (n *Float) UnmarshalJSON(b []byte) error {
+	var objMap map[string]*json.RawMessage
+	err := json.Unmarshal(b, &objMap)
+	if err != nil {
+		err = unmarshalFloat(b, n)
+
+		if err != nil {
+			return err
+		}
+	} else {
```

Comment:
This method should only ever be called by "UnmarshalType(tp MarshalType, b []byte) (Node, error)", which passes the encoding of the value in the b parameter. What is the purpose of the code in the "else" block below?

I think I know what the issue is. You are trying to make your top-level user-facing method
  func Unmarshal(r io.Reader, out Node) error
to work, when someone passes any concrete syntactic type for "out". 

This causes these "ugly" if/else statements here. The ugliness is actually the fact that you are checking for types in two different places in your code: once in every concrete type's unmarshal method, and once in the UnmarshalType method. This means that there is no one source of truth in the code and makes the code very prone to future bugs.

To fix this issue, change the top-level unmarshaling interface to:
func Unmarshal(r io.Reader) (Node, error)

Reply:
I'd say that I only check types in one place, when using UnmarshalJSON for Pairs. For the rest of types the json.Marshaller knows how to unmarshal the right type by itself, am I right?

And regarding the top-level interface, I did it this way because I wanted to use Decode under the hood, and I didn't know how to write a MarshalJSON for Node. In any case, would you mind if we discuss this sync? I think I may be missing something here. Thanks!